From: Mike Gilbert <floppym@gentoo.org>
Date: Thu, 5 Sep 2019 11:43:29 -0400
Subject: Fix expansion of bare tildes when HOME is unset

Use getpwuid to find the user name and home directory. Cache the user name for
future logins_tree lookups to avoid repeating the getpwuid call.

Fixes: https://github.com/att/ast/issues/1391
(cherry picked from commit c4f582db79876700d2ccf2cd98a39a205f39f48f)
---
 src/cmd/ksh93/sh/macro.c        | 13 +++++++++----
 src/cmd/ksh93/tests/builtins.sh |  8 ++++++++
 2 files changed, 17 insertions(+), 4 deletions(-)

diff --git a/src/cmd/ksh93/sh/macro.c b/src/cmd/ksh93/sh/macro.c
index 56fe707..7e641f9 100644
--- a/src/cmd/ksh93/sh/macro.c
+++ b/src/cmd/ksh93/sh/macro.c
@@ -2411,14 +2411,19 @@ static_fn void tilde_expand2(Shell_t *shp, int offset) {
 static_fn char *sh_tilde(Shell_t *shp, const char *string) {
     char *cp;
     int c;
-    struct passwd *pw;
+    struct passwd *pw = NULL;
     Namval_t *np = NULL;
     static Dt_t *logins_tree;
+    static char *username;
 
     if (*string++ != '~') return NULL;
     if ((c = *string) == 0) {
-        if (!(cp = nv_getval(sh_scoped(shp, HOME)))) cp = getlogin();
-        return cp;
+        if ((cp = nv_getval(sh_scoped(shp, HOME)))) return cp;
+        if (!username) {
+            if (!(pw = getpwuid(getuid()))) return NULL;
+            if (!(username = strdup(pw->pw_name))) return NULL;
+        }
+        string = username;
     }
     if ((c == '-' || c == '+') && string[1] == 0) {
         if (c == '+') {
@@ -2490,7 +2495,7 @@ static_fn char *sh_tilde(Shell_t *shp, const char *string) {
     }
 #endif  // __CYGWIN__
     if (logins_tree && (np = nv_search(string, logins_tree, 0))) return nv_getval(np);
-    if (!(pw = getpwnam(string))) return NULL;
+    if (!pw && !(pw = getpwnam(string))) return NULL;
 #if __CYGWIN__
 skip:
 #endif  // __CYGWIN__
diff --git a/src/cmd/ksh93/tests/builtins.sh b/src/cmd/ksh93/tests/builtins.sh
index a3830bf..55b1f4f 100644
--- a/src/cmd/ksh93/tests/builtins.sh
+++ b/src/cmd/ksh93/tests/builtins.sh
@@ -506,6 +506,14 @@ then
     rmdir "$pwd/f1"
 fi
 
+# =======
+# Verify `~` expansion works if `$HOME` is not set.
+# Regression test for https://github.com/att/ast/issues/1391
+expect="$(print ~$(id -un))"
+actual=$(unset HOME; $SHELL -c 'cd /; cd ~; pwd')
+[[ $actual == $expect ]] || log_error "bare ~ expansion with unset HOME" "$expect" "$actual"
+
+# =======
 TESTDIRSYMLINK="$TEST_DIR/testdirsymlink"
 ln -s "$TEST_DIR" "$TEST_DIR/testdirsymlink"
 
